pandaserver/dataservice/closer.py

```diff
@@ -0,0 +1,469 @@
+"""
+update dataset DB, and then close dataset and start Activator if needed
+
+"""
+
+import re
+import sys
+import datetime
+
+from typing import Dict, List
+
+from pandacommon.pandalogger.PandaLogger import PandaLogger
+from pandacommon.pandalogger.LogWrapper import LogWrapper
+
+from pandaserver.config import panda_config
+from pandaserver.dataservice import Notifier
+from pandaserver.dataservice.Activator import Activator
+from pandaserver.taskbuffer import EventServiceUtils
+
+# logger
+_logger = PandaLogger().getLogger("closer")
+
+
+def init_logger(p_logger: PandaLogger) -> None:
+    """
+    Redirect logging to parent as it doesn't work in nested threads
+
+    Args:
+        p_logger (PandaLogger): The parent logger.
+    """
+    # redirect logging to parent as it doesn't work in nested threads
+    global _logger
+    _logger = p_logger
+    Notifier.initLogger(_logger)
+
+
+class Closer:
+    """
+    Update dataset DB, and then close dataset and start Activator if needed.
+    The class updates output-related dataset records (_sub, output, and log) in
+    the database, closes _sub datasets if necessary, and activates downstream
+    runJobs if the output dataset is for buildJob (libDS, library dataset).
+    Activator changes job status from defined/assigned to activated.
+    """
+
+    # constructor
+    def __init__(self, taskBuffer, destination_dispatch_blocks: List[str], job, panda_ddm: bool = False,
+                 dataset_map: Dict = None) -> None:
+        """
+        Constructor
+
+        Args:
+            task_buffer: Task buffer.
+            destination_dispatch_blocks (List[str]): Destination Dispatch blocks.
+            job: Job.
+            panda_ddm (bool, optional): Panda DDM flag. Defaults to False.
+            dataset_map (Dict, optional): Dataset map. Defaults to {}.
+        """
+        self.task_buffer = taskBuffer
+        self.destination_dispatch_blocks = destination_dispatch_blocks
+        self.job = job
+        self.panda_id = job.PandaID
+        self.panda_ddm = panda_ddm
+        self.site_mapper = None
+        self.dataset_map = dataset_map if dataset_map is not None else {}
+        self.all_subscription_finished = None
+        self.first_indv_dataset = True
+        self.using_merger = False
```

Comment:
needed?

Reply:
@tmaeno could you please comment?

Reply:
first_indv_dataset and using_merger are included in the elif block that can be deleted.